include/openssl/hpke.h

```diff
@@ -0,0 +1,139 @@
+
+# define OSSL_HPKE_KDF_ID_RESERVED         0x0000 /**< not used */
+# define OSSL_HPKE_KDF_ID_HKDF_SHA256      0x0001 /**< HKDF-SHA256 */
+# define OSSL_HPKE_KDF_ID_HKDF_SHA384      0x0002 /**< HKDF-SHA384 */
+# define OSSL_HPKE_KDF_ID_HKDF_SHA512      0x0003 /**< HKDF-SHA512 */
+
+# define OSSL_HPKE_AEAD_ID_RESERVED        0x0000 /**< not used */
+# define OSSL_HPKE_AEAD_ID_AES_GCM_128     0x0001 /**< AES-GCM-128 */
+# define OSSL_HPKE_AEAD_ID_AES_GCM_256     0x0002 /**< AES-GCM-256 */
+# define OSSL_HPKE_AEAD_ID_CHACHA_POLY1305 0x0003 /**< Chacha20-Poly1305 */
+# define OSSL_HPKE_AEAD_ID_EXPORTONLY      0xFFFF /**< export-only fake ID */
+
+/* strings for suite components - ideally these'd be defined elsewhere */
+# define OSSL_HPKE_KEMSTR_P256        "P-256"              /**< KEM id 0x10 */
+# define OSSL_HPKE_KEMSTR_P384        "P-384"              /**< KEM id 0x11 */
+# define OSSL_HPKE_KEMSTR_P521        "P-521"              /**< KEM id 0x12 */
+# define OSSL_HPKE_KEMSTR_X25519      SN_X25519            /**< KEM id 0x20 */
+# define OSSL_HPKE_KEMSTR_X448        SN_X448              /**< KEM id 0x21 */
+# define OSSL_HPKE_KDFSTR_256         "hkdf-sha256"        /**< KDF id 1 */
+# define OSSL_HPKE_KDFSTR_384         "hkdf-sha384"        /**< KDF id 2 */
+# define OSSL_HPKE_KDFSTR_512         "hkdf-sha512"        /**< KDF id 3 */
+# define OSSL_HPKE_AEADSTR_AES128GCM  LN_aes_128_gcm       /**< AEAD id 1 */
+# define OSSL_HPKE_AEADSTR_AES256GCM  LN_aes_256_gcm       /**< AEAD id 2 */
+# define OSSL_HPKE_AEADSTR_CP         LN_chacha20_poly1305 /**< AEAD id 3 */
+# define OSSL_HPKE_AEADSTR_EXP        "exporter"           /**< AEAD id 0xff */
+
+typedef struct {
+    uint16_t    kem_id; /**< Key Encapsulation Method id */
+    uint16_t    kdf_id; /**< Key Derivation Function id */
+    uint16_t    aead_id; /**< AEAD alg id */
+} OSSL_HPKE_SUITE;
+
+/**
+ * Suite constants, use this like:
+ *          OSSL_HPKE_SUITE myvar = OSSL_HPKE_SUITE_DEFAULT;
+ */
+# define OSSL_HPKE_SUITE_DEFAULT \
+    {\
+        OSSL_HPKE_KEM_ID_X25519, \
+        OSSL_HPKE_KDF_ID_HKDF_SHA256, \
+        OSSL_HPKE_AEAD_ID_AES_GCM_128 \
+    }
+
+typedef struct ossl_hpke_ctx_st OSSL_HPKE_CTX;
+
+OSSL_HPKE_CTX *OSSL_HPKE_CTX_new(int mode, OSSL_HPKE_SUITE suite,
+                                 OSSL_LIB_CTX *libctx, const char *propq);
+void OSSL_HPKE_CTX_free(OSSL_HPKE_CTX *ctx);
+
+int OSSL_HPKE_encap(OSSL_HPKE_CTX *ctx,
+                    unsigned char *enc, size_t *enclen,
+                    const unsigned char *pub, size_t publen,
+                    const unsigned char *info, size_t infolen);
+int OSSL_HPKE_seal(OSSL_HPKE_CTX *ctx,
+                   unsigned char *ct, size_t *ctlen,
+                   const unsigned char *aad, size_t aadlen,
+                   const unsigned char *pt, size_t ptlen);
+
+int OSSL_HPKE_keygen(OSSL_HPKE_SUITE suite,
+                     const unsigned char *ikm, size_t ikmlen,
```

Comment:
ikm is optional so move it down?

Reply:
sure, if that better matches usual patterns